[FSDP][state_dict][4/N] Move the core logic of summon full parameters to _unshard_params_utils.py

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88638
* #88637
* __->__ #88636
* #88635
* #88481
* #87900

**What**
`_summon_full_parameters` is required for state_dict. To enable composable FSDP state_dict, `_summon_full_params` must be accessible without FullyShardedDataParall. This PR move the core logic of `_summon_full_params` to `_unshard_params_utils`.

fegin committed:
[FSDP][state_dict][4/N] Move the core logic of summon full parameters to _unshard_params_utils.py

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][state_dict][4/N] Move the core logic of summon full parameters to _unshard_params_utils.py"


**What**
`_summon_full_parameters` is required for state_dict. To enable composable FSDP state_dict, `_summon_full_params` must be accessible without FullyShardedDataParall. This PR move the core logic of `_summon_full_params` to `_unshard_params_utils`.


[ghstack-poisoned]

fegin committed:
Update on "[FSDP][state_dict][4/N] Move the core logic of summon full parameters to _unshard_params_utils.py"


**What**
`_summon_full_parameters` is required for state_dict. To enable composable FSDP state_dict, `_summon_full_params` must be accessible without FullyShardedDataParall. This PR move the core logic of `_summon_full_params` to `_unshard_params_utils`.


[ghstack-poisoned]

fegin committed:
Update on "[FSDP][state_dict][4/N] Move the core logic of summon full parameters to _unshard_params_utils.py"


**What**
`_summon_full_parameters` is required for state_dict. To enable composable FSDP state_dict, `_summon_full_params` must be accessible without FullyShardedDataParall. This PR move the core logic of `_summon_full_params` to `_unshard_params_utils`.


[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_unshard_param_utils.py:
```diff
@@ -0,0 +1,254 @@
+                )
+                existing_grad[: grad_shard.numel()].copy_(grad_shard)
+
+
```
nit: Same comment as previous PR. I am wondering if you prefer to have the arguments in the current order or reversed. I am not sure if there may be confusion, but I do not feel strongly either way.

awgu commented on torch/distributed/fsdp/_unshard_param_utils.py:
```diff
@@ -0,0 +1,254 @@
+    """
+    handles = _module_handles(module, state)
+    if _has_fsdp_params(module, state):
+        # TODO: figure out the case for the composable APIs.
```
It seems that we need to be able to get the module from the `_State` object? I wonder if this is a problem and not possible with the current composable design. (Maybe we are only allowed to modify the `_State` object embedded on the module but not the module itself.)

cc: @zhaojuanmao @mrshenli

awgu commented on torch/distributed/fsdp/_state_dict_utils.py:
```diff
@@ -338,6 +331,7 @@ def _full_pre_load_state_dict_hook(
     state_dict: Dict[str, Any],
     prefix: str,
 ) -> None:
+    _lazy_init(fsdp_state, module)
```
Could you remind me: Is there any functional change that happens by moving the `_lazy_init()` call into the pre-hook?

In my current understanding, there is no change (since `_lazy_init()` is a no-op if already initialized and since the hooks will run top-down as desired).

fegin replied:
Yes, if `pre_state_dict_hook` is called from root to leaf, there is no functional change.

awgu reviewed (approved):
LGTM! We should follow-up about how to register the `FlatParameter` for the composable case.

fegin committed:
Update on "[FSDP][state_dict][4/N] Move the core logic of summon full parameters to _unshard_params_utils.py"


**What**
`_summon_full_parameters` is required for state_dict. To enable composable FSDP state_dict, `_summon_full_params` must be accessible without FullyShardedDataParall. This PR move the core logic of `_summon_full_params` to `_unshard_params_utils`.


[ghstack-poisoned]

fegin committed:
Update on "[FSDP][state_dict][4/N] Move the core logic of summon full parameters to _unshard_params_utils.py"


**What**
`_summon_full_parameters` is required for state_dict. To enable composable FSDP state_dict, `_summon_full_params` must be accessible without FullyShardedDataParall. This PR move the core logic of `_summon_full_params` to `_unshard_params_utils`.


[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fegin commented:
@pytorchbot merge -f "The failing test is not related."
